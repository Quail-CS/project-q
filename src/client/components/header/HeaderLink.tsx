/**
 * @module HeaderLink.tsx
 * @description Application Header Link Presentation Component
 * UNIT TEST COVERAGE - 100%
 */

import * as React from 'react';
import { Link } from 'react-router-dom';

export interface HeaderLinkProps {
  className: string,
  link: string,
  linkKey: string,
  linkText: string,
}

const HeaderLink = (props: HeaderLinkProps): JSX.Element => {
  const { className, linkKey, link, linkText } = props;
  return (
    <li className={className} key={linkKey}><Link to={link}>{linkText}</Link></li>
  )
}

export default HeaderLink;
