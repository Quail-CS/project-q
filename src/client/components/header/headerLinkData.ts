/**
 * @module headerLinkData.ts
 * @description List of Header Links
 */

interface IHeaderLinkData {
  [linkName: string]: {
    className: string,
    link: string,
    linkKey: string,
    linkText: string,
    modal: boolean,
    onClick?: () => any,
  }
}

export const headerLinkData: IHeaderLinkData = {
  logout: {
    className: 'header-right',
    link: '/account/logout',
    linkKey: 'logout',
    linkText: 'LOGOUT',
    modal: false,
  },
  restart: {
    className: 'header-right',
    link: '/account/logout',
    linkKey: 'restart',
    linkText: 'RESET',
    modal: true,
  },
  home: {
    className: 'header-right',
    link: '/',
    linkKey: 'home',
    linkText: 'HOME',
    modal: false,
  },
  register: {
    className: 'header-right',
    link: '/account/register',
    linkKey: 'register',
    linkText: 'REGISTER',
    modal: false,
  },
  login: {
    className: 'header-right',
    link: '/account/login',
    linkKey: 'login',
    linkText: 'LOGIN',
    modal: false,
  },
};
