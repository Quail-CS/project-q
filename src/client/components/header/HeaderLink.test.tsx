import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as Renderer from 'react-test-renderer';

import HeaderLink from './HeaderLink';
import { MemoryRouter } from 'react-router';
import { headerLinkData } from './headerLinkData';
import { Link } from 'react-router-dom';

Enzyme.configure({ adapter: new Adapter() });

const props = {
  className: headerLinkData.login.className,
  linkKey: headerLinkData.login.linkKey,
  linkRoute: headerLinkData.login.linkRoute,
}

describe('HeaderLink Component', () => {
  it('matches the snapshot', () => {
    const tree: Renderer.ReactTestRendererJSON = Renderer.create(
      <MemoryRouter>
        <HeaderLink {...props} linkRoute={headerLinkData.login.linkRoute as string}>{headerLinkData.login.linkText}</HeaderLink>
      </MemoryRouter>
    ).toJSON() as Renderer.ReactTestRendererJSON;
    expect(tree).toMatchSnapshot();
  })
  it('renders a <li> with key equal to linkKey prop', () => {
    const headerLinkWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<HeaderLink {...props} linkRoute={headerLinkData.login.linkRoute as string}>{headerLinkData.login.linkText}</HeaderLink>)
    expect(headerLinkWrapper.type()).toEqual('li');
    expect(headerLinkWrapper.key()).toEqual(props.linkKey);
  })
  it('contains a Link to the link prop with text equal to the text prop', () => {
    const headerLinkWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<HeaderLink {...props} linkRoute={headerLinkData.login.linkRoute as string}>{headerLinkData.login.linkText}</HeaderLink>)
    expect(headerLinkWrapper.find(Link)).toHaveLength(1);
    expect(headerLinkWrapper.find(Link).props()).toHaveProperty('children', headerLinkData.login.linkText)
    expect(headerLinkWrapper.find(Link).props()).toHaveProperty('to', props.linkRoute)
  })
});
