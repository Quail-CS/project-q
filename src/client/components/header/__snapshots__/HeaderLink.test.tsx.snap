// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderLink Component matches the snapshot 1`] = `
<li
  className="header-right"
>
  <a
    href="/account/login"
    onClick={[Function]}
  >
    LOGIN
  </a>
</li>
`;
